File: app/src/main/java/io/github/ovso/righttoknow/violator/ViolatorAdapter.java
package io.github.ovso.righttoknow.violator;

import android.view.View;
import android.widget.TextView;
import butterknife.BindView;
import io.github.ovso.righttoknow.R;
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView;
import io.github.ovso.righttoknow.framework.adapter.BaseRecyclerAdapter;
import io.github.ovso.righttoknow.framework.adapter.OnRecyclerItemClickListener;
import io.github.ovso.righttoknow.violator.model.Violator;
import java.util.ArrayList;
import java.util.List;
import lombok.Setter;

/**
 * Created by jaeho on 2017. 8. 3
 */

public class ViolatorAdapter extends BaseRecyclerAdapter
    implements BaseAdapterView, ViolatorAdapterDataModel<Violator> {

  private List<Violator> items = new ArrayList<>();

  @Override protected BaseViewHolder createViewHolder(View view, int viewType) {
    return new ViolatorViewHolder(view);
  }

  @Override public int getLayoutRes(int viewType) {
    return R.layout.fragment_violator_item;
  }

  @Override public void onBindViewHolder(BaseViewHolder baseHolder, int position) {
    if (baseHolder instanceof ViolatorViewHolder) {
      ViolatorViewHolder holder = (ViolatorViewHolder) baseHolder;
      final Violator violator = items.get(position);
      holder.orderTextview.setText(String.valueOf(violator.getOrder()));
      holder.sidoTextView.setText(violator.getSido());
      holder.sigunguTextView.setText(violator.getSigungu());
      holder.nameTextView.setText(violator.getName());
      holder.vioFacTextView.setText(violator.getFac_name());
      holder.historyTextView.setText(violator.getHistory());
      holder.itemView.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View view) {
          onRecyclerItemClickListener.onItemClick(violator);
        }
      });
    }
  }

  @Override public int getItemCount() {
    return getSize();
  }

  @Override public void refresh() {
    notifyDataSetChanged();
  }

  @Override public void add(Violator item) {
    items.add(item);
  }

  @Override public void addAll(List<Violator> items) {
    this.items.addAll(items);
  }

  @Override public Violator remove(int position) {
    return items.remove(position);
  }

  @Override public Violator getItem(int position) {
    return items.get(position);
  }

  @Override public void add(int index, Violator item) {
    items.add(index, item);
  }

  @Override public int getSize() {
    return items.size();
  }

  @Setter private OnRecyclerItemClickListener<Violator> onRecyclerItemClickListener;

  final static class ViolatorViewHolder extends BaseRecyclerAdapter.BaseViewHolder {
    @BindView(R.id.order_textview) TextView orderTextview;
    @BindView(R.id.sido_textview) TextView sidoTextView;
    @BindView(R.id.sigungu_textview) TextView sigunguTextView;
    @BindView(R.id.name_textview) TextView nameTextView;
    @BindView(R.id.vio_fac_textview) TextView vioFacTextView;
    @BindView(R.id.history_textview) TextView historyTextView;

    public ViolatorViewHolder(View itemView) {
      super(itemView);
    }
  }

  @Override public void clear() {
    items.clear();
  }
}

File: app/src/main/java/io/github/ovso/righttoknow/ui/main/violator/ViolatorAdapter.kt
package io.github.ovso.righttoknow.ui.main.violator

import android.view.View
import io.github.ovso.righttoknow.R
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView
import io.github.ovso.righttoknow.framework.adapter.BaseRecyclerAdapter
import io.github.ovso.righttoknow.framework.adapter.OnRecyclerItemClickListener
import io.github.ovso.righttoknow.ui.main.violator.model.Violator
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.fragment_violator_item.*
import java.util.*

class ViolatorAdapter : BaseRecyclerAdapter(), BaseAdapterView,
  ViolatorAdapterDataModel<Violator> {
  private val items: MutableList<Violator> = ArrayList()
  override fun createViewHolder(view: View, viewType: Int): BaseViewHolder {
    return ViolatorViewHolder(view)
  }

  override fun getLayoutRes(viewType: Int): Int {
    return R.layout.fragment_violator_item
  }

  override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
    if (holder is ViolatorViewHolder) {
      holder.onBindViewHolder(getItem(position))
      holder.itemView.setOnClickListener {
        onRecyclerItemClickListener.onItemClick(
          getItem(
            position
          )
        )
      }
    }
  }

  override fun getItemCount(): Int {
    return size
  }

  override fun refresh() {
    notifyDataSetChanged()
  }

  override fun add(item: Violator) {
    items.add(item)
  }

  override fun addAll(items: List<Violator>) {
    this.items.addAll(items)
  }

  override fun remove(position: Int): Violator {
    return items.removeAt(position)
  }

  override fun getItem(position: Int): Violator {
    return items[position]
  }

  override fun add(index: Int, item: Violator) {
    items.add(index, item)
  }

  override val size: Int
    get() = items.size

  lateinit var onRecyclerItemClickListener: OnRecyclerItemClickListener<Violator>

  internal class ViolatorViewHolder(override val containerView: View?) :
    BaseViewHolder(containerView!!), LayoutContainer {
    fun onBindViewHolder(item: Violator) {
      order_textview.text = item.order.toString()
      sido_textview.text = item.sido
      sigungu_textview.text = item.sigungu
      name_textview.text = item.name
      vio_fac_textview.text = item.fac_name
      history_textview.text = item.history
    }
  }

  override fun clear() {
    items.clear()
  }
}
